### PR TITLE
Clickhouse refactor connection code and set secure option

### PR DESCRIPTION
**Pull request by gloomweaver**
- Refactor setting up connection to clickhouse
- Support `clickhouse_secure` parameter for setting up ssl connection
- Tested against clickhouse cloud instance

### CODE REVIEW

File: crates/db_connection_pool/src/clickhousepool.rs [hunk 1]
```diff
@@ -109,6 +81,84 @@ impl ClickhouseConnectionPool {
...
+    if let Some(clickhouse_connection_string) = get_secret_or_param(
+        params,
+        secret,
+        "clickhouse_connection_string_key",
+        "clickhouse_connection_string",
+    ) {
+        let mut new_options = Options::from_str(&clickhouse_connection_string)
+            .context(InvalidConnectionStringSnafu)?;
+        if !clickhouse_connection_string.contains("connection_timeout") {
+            // Default timeouf of 500ms is not enough in some cases.
```
Comment (digadeesh) [inline crates/db_connection_pool/src/clickhousepool.rs]:
```suggestion
            // Default timeout of 500ms is not enough in some cases.
```

Comment (digadeesh) [reply crates/db_connection_pool/src/clickhousepool.rs]: Was 5 seconds picked at random?

File: crates/db_connection_pool/src/clickhousepool.rs [hunk 2]
```diff
@@ -109,6 +81,84 @@ impl ClickhouseConnectionPool {
...
+        }
+        let connection_string = format!(
+            "tcp://{}:{}@{}:{}/{}",
+            connection_opts[0],
+            connection_opts[1],
+            connection_opts[2],
+            connection_opts[3],
+            connection_opts[4]
+        );
+        // Default timeouf of 500ms is not enough
```
Comment (digadeesh) [inline crates/db_connection_pool/src/clickhousepool.rs]:
```suggestion
        // Default timeout of 500ms is not enough
```

Comment (digadeesh) [reply crates/db_connection_pool/src/clickhousepool.rs]: Also, copy/paste code.  At minimum this should be a constant, but it's a little peculiar to format it to a string, just so you can parse it back into options.  Can we refactor this to be a little cleaner?

File: crates/db_connection_pool/src/clickhousepool.rs (hunk 1 above)
Comment (gloomweaver) [reply crates/db_connection_pool/src/clickhousepool.rs]: Felt like a good default value.
Will change to default specified by clickhouse https://clickhouse.com/docs/en/operations/settings/settings#connect-timeout-receive-timeout-send-timeout
Also will create a fix for clickhouse-rs crate

File: crates/db_connection_pool/src/clickhousepool.rs (hunk 2 above)
Comment (gloomweaver) [reply crates/db_connection_pool/src/clickhousepool.rs]: It's a quirk of the clickhouse-rs crate. Options struct doesn't expose all necessary methods to construct manually.
Will open up a pr to clickhouse-rs to fix those API problems.

File: crates/db_connection_pool/src/clickhousepool.rs [hunk 3]
```diff
@@ -109,6 +92,91 @@ impl ClickhouseConnectionPool {
...
+            .fail()?,
+        }
+        let connection_string = format!(
+            "tcp://{}:{}@{}:{}/{}",
+            connection_opts[0],
+            connection_opts[1],
+            connection_opts[2],
+            connection_opts[3],
+            connection_opts[4]
+        );
```
Comment (ewgenius) [inline crates/db_connection_pool/src/clickhousepool.rs]: This doesn't look durable enough, and it is hard to follow which parameter goes in each position. I would prefer to explicitly retrieve each parameter rather than relying on loops.

File: crates/db_connection_pool/src/clickhousepool.rs [hunk 4]
```diff
@@ -42,8 +47,20 @@ pub enum Error {
         source: clickhouse_rs::errors::Error,
     },
 
-    #[snafu(display("Invalid parameter: {parameter_name}"))]
-    InvalidParameterError { parameter_name: String },
+    #[snafu(display("ParametersEmptyError"))]
+    ParametersEmptyError {},
+
+    #[snafu(display("Required parameter was not provided: {parameter_name}"))]
+    NotEnoughParametersForConnection { parameter_name: String },
```
Comment (ewgenius) [inline crates/db_connection_pool/src/clickhousepool.rs]:
```suggestion
    #[snafu(display("Missing required parameter: {parameter_name}"))]
    MissingRequiredParameterForConnection { parameter_name: String },
```

File: crates/db_connection_pool/src/clickhousepool.rs (hunk 3 above)
Comment (ewgenius) [reply crates/db_connection_pool/src/clickhousepool.rs]: Something like

```rust

let user = params.get("clickhouse_user")
     .ok_or_else(|| MissingRequiredParameter { name: "clickhouse_user" });

let host = params.get("clickhouse_host")
     .ok_or_else(|| MissingRequiredParameter { name: "clickhouse_host" });

...

let connection_string = format!("tcp://{user}:{host}...");

```

File: crates/db_connection_pool/src/clickhousepool.rs [hunk 5]
```diff
@@ -109,6 +92,91 @@ impl ClickhouseConnectionPool {
...
+    let secure = params
+        .get("clickhouse_secure")
+        .map(|s| s.parse::<bool>())
+        .transpose()
+        .context(InvalidSecureParameterValueSnafu {
+            parameter_name: "clickhouse_secure".to_string(),
+        })?;
+    options = options.secure(secure.unwrap_or(true));
+
+    if let Some(clickhouse_sslrootcert) = params.get("clickhouse_sslrootcert") {
```
Comment (ewgenius) [inline crates/db_connection_pool/src/clickhousepool.rs]: Where is `clickhouse_sslrootcert` used? Should it be passed to options?

File: crates/db_connection_pool/src/clickhousepool.rs [hunk 6]
```diff
@@ -109,6 +92,91 @@ impl ClickhouseConnectionPool {
...
+            connection_opts[0],
+            connection_opts[1],
+            connection_opts[2],
+            connection_opts[3],
+            connection_opts[4]
+        );
+        // Default timeout of 500ms is not enough
+        let new_options = Options::from_str(&connection_string)
+            .context(InvalidConnectionStringSnafu)?
+            .connection_timeout(DEFAULT_CONNECTION_TIMEOUT);
```
Comment (ewgenius) [inline crates/db_connection_pool/src/clickhousepool.rs:150]: I think that should be also configurable through options - `params.clickhouse_connection_timeout`

File: crates/db_connection_pool/src/clickhousepool.rs (hunk 3 above)
Comment (gloomweaver) [reply crates/db_connection_pool/src/clickhousepool.rs]: Fixed

File: crates/db_connection_pool/src/clickhousepool.rs (hunk 5 above)
Comment (gloomweaver) [reply crates/db_connection_pool/src/clickhousepool.rs]: Removed, currently there is a problem this crate.
It doesn't expose the required code.

File: crates/db_connection_pool/src/clickhousepool.rs (hunk 6 above)
Comment (gloomweaver) [reply crates/db_connection_pool/src/clickhousepool.rs:150]: Added

File: crates/db_connection_pool/src/clickhousepool.rs [hunk 7]
```diff
@@ -42,8 +47,20 @@ pub enum Error {
         source: clickhouse_rs::errors::Error,
     },
 
-    #[snafu(display("Invalid parameter: {parameter_name}"))]
-    InvalidParameterError { parameter_name: String },
+    #[snafu(display("ParametersEmptyError"))]
```
Comment (phillipleblanc) [inline crates/db_connection_pool/src/clickhousepool.rs]:
```suggestion
    #[snafu(display("No parameters specified"))]
```

**Review by ewgenius — APPROVED** [review]
